### PR TITLE
Move all vars to foo/defaults.sh and `set -Eeuo pipefail` globally

### LINKED ISSUE #77: Define variables needed by each module
We could come up with a strategy to define in one place the variables needed per-module, which than could be re-used to generate docs automatically. e.g. an `env.sh` for each module which contains the defaults.

Could be done at the same time when considering: https://github.com/SUSE/catapult/issues/72

### PR DESCRIPTION
**Pull request by viccuad**
This PR moves all env vars used as options to:
* `backend/foo/defaults.sh`
* `modules/foo/defaults.sh`
* although some global ones stay in `include/common.sh`

This unblocks concatenating all defaults and have documentation as code.

This allows to use `set -Eeuo pipefail` at the end of `include/common.sh`, which makes it the default for everything that includes it, making everything substantially more robust.

Includes some bug fixes found when discovering unbound variables, and scripts erroring, thanks to having `set -eu` globally.

Adds `kubectl` and `helm` mocks to pass unittest when on `set -e`.

Closes #77
Unblocks #72, #78

### CODE REVIEW

**Comment by viccuad** [comment]
it works quite well, but it has yet some edge cases that need to be polished. See TODO and FIXME added on the commits.

**Comment by viccuad** [comment]
Unit tests fail because we call scf-gen-config without having a cluster. And there we try to read the configmap of an unexistent cluster. This is an error that we see now, because now we default to `set -Eeuo pipefail`.

**Comment by viccuad** [comment]
Ready for review. Ended up being a bit of refactor in some tasks, but I am glad about setting `set -uo pipefail` globally, it is way more robust now.

I needed to deactivate testscfChart though.

**Comment by viccuad** [comment]
Once the review is ok, I can happily rebase

**Comment by viccuad** [comment]
Thanks to this, when running `make buildir` (almost always), all default options will be concatenated automatically and saved into buildir/defaults.sh.

Though that file isn't yet used to create a defaults.json, or read for modifications.

Eg:

```
#!/usr/bin/env bash
# DEFAULT VALUES. DO NOT CHANGE THIS FILE

# Global options
################

export SCF_REPO="${SCF_REPO:-https://github.com/SUSE/scf}"
export SCF_BRANCH="${SCF_BRANCH:-develop}"
export DOCKER_REGISTRY="${DOCKER_REGISTRY:-registry.suse.com}"
export DOCKER_ORG="${DOCKER_ORG:-cap}"

export DEFAULT_STACK="${DEFAULT_STACK:-from_chart}" # from_chart, sle15, sle12, cfslinuxfs2, cfslinuxfs3
export MAGICDNS="${MAGICDNS:-nip.io}"
export ENABLE_EIRINI="${ENABLE_EIRINI:-true}"
export EKCP_PROXY="${EKCP_PROXY:-}"
export KUBEPROXY_PORT="${KUBEPROXY_PORT:-2224}"
export QUIET_OUTPUT="${QUIET_OUTPUT:-false}"

# Global private defaults (not shown on output)
###############################################

export DOCKER_USERNAME="${DOCKER_USERNAME:-}"
export DOCKER_PASSWORD="${DOCKER_PASSWORD:-}"
export CLUSTER_PASSWORD="${CLUSTER_PASSWORD:-password}"

# Caasp4os options
##################

CAASP_VER=${CAASP_VER:-"product"}
STACK=${STACK:-"$(whoami)-caasp4-${CAASP_VER::3}-$CLUSTER_NAME"}

# EKCP options
##############

EKCP_HOST=${EKCP_HOST:-}
FORCE_DELETE=${FORCE_DELETE:-"false"}

# EKS options
#############

EKS_LOCATION="${EKS_LOCATION:-us-west-2}"
EKS_KEYPAIR="${EKS_KEYPAIR:-$(whoami)-terraform}"
EKS_VERS="${EKS_VERS:-1.14}"
EKS_CLUSTER_LABEL=${EKS_CLUSTER_LABEL:-{key = \"$(whoami)-eks-cluster\"}}

# GKE options
#############

GKE_PROJECT="${GKE_PROJECT:-suse-css-platform}"
GKE_LOCATION="${GKE_LOCATION:-europe-west4-a}"
GKE_CLUSTER_NAME="${GKE_CLUSTER_NAME:-$(whoami)-cap}"
GKE_CRED_JSON="${GKE_CRED_JSON:-}"

# Imported backend options
##########################

KUBECFG="${KUBECFG-}" # full path to kubeconfig

# Kind backend options
######################

KIND_VERSION="${KIND_VERSION:-0.6.1}"
KIND_APIVERSION="${KIND_APIVERSION:-kind.sigs.k8s.io/v1alpha3}"

# Minikube backend options
##########################

MINIKUBE_RUNTIME=${MINIKUBE_RUNTIME:-docker} # docker, containerd, cri-o

# Dependencies options
######################

HELM_VERSION="${HELM_VERSION:-v2.16.1}"

# Experimental module options
#############################

EIRINI_RELEASE_REPO="${EIRINI_RELEASE_REPO:-https://github.com/mudler/eirini-release}"
EIRINI_RELEASE_CHECKOUT="${EIRINI_RELEASE_CHECKOUT:-eirini_logging}"

EIRINIFS=${EIRINIFS:-https://github.com/os-fun/eirinifs.git}
EIRINISSH=${EIRINISSH:-https://github.com/SUSE/eirini-ssh}

# Extra module options
######################

KWT_VERSION="${KWT_VERSION:-v0.0.5}"
KWT_OS_TYPE="${KWT_OS_TYPE:-linux-amd64}"

TMPDIR="${TMPDIR:-/tmp}" # required to be able to mount configs in deployments containers (dind mount )

# Metrics options
#################

METRICS_CHART="${METRICS_CHART:-latest}"

# SCF options
#############

CHART_URL="${CHART_URL:-}" # FIXME deprecated, used in SCF_CHART
SCF_CHART="${SCF_CHART:-$CHART_URL}"
STRATOS_CHART="${STRATOS_CHART:-}"

SCF_HELM_VERSION="${SCF_HELM_VERSION:-}"
SCF_OPERATOR="${SCF_OPERATOR:-false}"
OPERATOR_CHART_URL="${OPERATOR_CHART_URL:-latest}"

GARDEN_ROOTFS_DRIVER="${GARDEN_ROOTFS_DRIVER:-btrfs}"
DIEGO_SIZING="${DIEGO_SIZING:-$SIZING}"
STORAGECLASS="${STORAGECLASS:-persistent}"
AUTOSCALER="${AUTOSCALER:-false}"

EMBEDDED_UAA="${EMBEDDED_UAA:-false}"

HA="${HA:-false}"
if [ "$HA" = "true" ]; then
    SIZING="${SIZING:-2}"
else
    SIZING="${SIZING:-1}"
fi

UAA_UPGRADE="${UAA_UPGRADE:-true}"

# stratos module options
########################

STRATOS_CHART="${STRATOS_CHART:-latest}"

# SMOKES option
###############

SMOKES_REPO="${SMOKES_REPO:-https://github.com/cloudfoundry/cf-smoke-tests}"

# CATS options
##############

CATS_REPO="${CATS_REPO:-https://github.com/cloudfoundry/cf-acceptance-tests}"

# BRATS options
###############

PROXY_SCHEME="${PROXY_SCHEME:-http}"
BRATS_CF_USERNAME="${BRATS_CF_USERNAME:-admin}"
BRATS_CF_PASSWORD="${BRATS_CF_PASSWORD:-$CLUSTER_PASSWORD}"
PROXY_PORT="${PROXY_PORT:-9002}"
PROXY_USERNAME="${PROXY_USERNAME:-username}"
PROXY_PASSWORD="${PROXY_PASSWORD:-password}"
BRATS_TEST_SUITE="${BRATS_TEST_SUITE:-brats}"
CF_STACK="${CF_STACK:-sle15}"
GINKGO_ATTEMPTS="${GINKGO_ATTEMPTS:-3}"

BRATS_BUILDPACK="${BRATS_BUILDPACK}"
BRATS_BUILDPACK_URL="${BRATS_BUILDPACK_URL}"
BRATS_BUILDPACK_VERSION="${BRATS_BUILDPACK_VERSION}"

# Sample app options
####################

SAMPLE_APP_REPO="${SAMPLE_APP_REPO:-https://github.com/cloudfoundry-samples/cf-sample-app-nodejs}"

```

File: modules/experimental/eirini_release.sh [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
 
-if [ "$KIND_VERSION" != "v0.4.0" ]; then
- err "We only support Kind version 0.4.0 for now"
-
- exit 1
+. ./defaults.sh
+. ../../include/common.sh
+. .envrc
 
+if [ -f "$BUILD_DIR"/bin/kind ] && [ "$("BUILD_DIR"/bin/kind version)" != "v0.4.0" ]; then
```
Comment (viccuad) [inline modules/experimental/eirini_release.sh]: @mudler is this still needed?

Comment (mudler) [reply modules/experimental/eirini_release.sh]: the only version tested there was `v0.4.0`

Comment (mudler) [reply modules/experimental/eirini_release.sh]: but at the times, it was to check that you were running the latest (and greatest)

Comment (viccuad) [reply modules/experimental/eirini_release.sh]: So it could be removed actually
